### PR TITLE
L1T Phase-2: add tracking jets

### PR DESCRIPTION
**Pull request by cecilecaillol**
#### PR description:

This PR is for the introduction of Tracking Jets Producer , which clusters L1 tracks that are matched with Gen particles to create jets. The clustering algorithm that was used is the "FastJet". Essentially, is the optimal performance we can get in L1 jets, and be viewed as the ceiling for the phase2 track jet algo. This can be used in the future to assess performance of new algo / Fast Jet. More can be found in the GTT presentation: https://indico.cern.ch/event/1028525/contributions/4324252/attachments/2228034/3774593/Performance_TrkJet.pdf

#### PR validation:

Rebase of https://github.com/cms-l1t-offline/cmssw/pull/914

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Rebase of https://github.com/cms-l1t-offline/cmssw/pull/914

### CODE REVIEW

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
+l1

**Comment by srimanob** [comment]
Hi @cecilecaillol 

Not related to this PR, but because it compiles L1Trigger/L1TTrackMatch, there are several warning I assume it should be checked/fixed.
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c3465f/16224/llvm-analysis/runStaticChecks.log

FYI @qliphy @silviodonato

File: L1Trigger/L1TTrackMatch/plugins/L1FastTrackingJetProducer.cc [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+
+  edm::Handle<std::vector<l1t::Vertex>> L1VertexHandle;
+  iEvent.getByToken(pvToken_, L1VertexHandle);
+
+  fastjet::JetDefinition jet_def(fastjet::antikt_algorithm, coneSize_);
+  std::vector<fastjet::PseudoJet> JetInputs;
+
+  float recoVtx = L1VertexHandle->begin()->z0();
+  unsigned int this_l1track = 0;
+  for (iterL1Track = TTTrackHandle->begin(); iterL1Track != TTTrackHandle->end(); iterL1Track++) {
```
Comment (sroychow) [inline L1Trigger/L1TTrackMatch/plugins/L1FastTrackingJetProducer.cc:149]: Suggest range based loops here and elsewhere in the code.

File: L1Trigger/L1TTrackMatch/plugins/L1FastTrackingJetProducer.cc [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+      continue;
+    if (trk_pt < trkPtMin_)
+      continue;
+    if (trk_nstub < trkNStubMin_)
+      continue;
+    if (trk_chi2dof > trkChi2dofMax_)
+      continue;
+    if (trk_bendchi2 > trkBendChi2Max_)
+      continue;
+    if (doTightChi2_ && (trk_pt > 20.0 && trk_chi2dof > 5.0))
```
Comment (sroychow) [inline L1Trigger/L1TTrackMatch/plugins/L1FastTrackingJetProducer.cc:171]: Can't you pass 20. and 5 as parameters from the config?

File: L1Trigger/L1TTrackMatch/plugins/L1FastTrackingJetProducer.cc [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    this_l1track++;
+    float trk_pt = iterL1Track->momentum().perp();
+    float trk_z0 = iterL1Track->z0();
+    float trk_chi2dof = iterL1Track->chi2Red();
+    float trk_bendchi2 = iterL1Track->stubPtConsistency();
+    std::vector<edm::Ref<edmNew::DetSetVector<TTStub<Ref_Phase2TrackerDigi_>>, TTStub<Ref_Phase2TrackerDigi_>>>
+        theStubs = iterL1Track->getStubRefs();
+    int trk_nstub = (int)theStubs.size();
+
+    if (fabs(trk_z0) > trkZMax_)
```
Comment (sroychow) [inline L1Trigger/L1TTrackMatch/plugins/L1FastTrackingJetProducer.cc:159]: std::abs

File: L1Trigger/L1TTrackMatch/plugins/L1FastTrackingJetProducer.cc [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+      continue;
+    if (fabs(iterL1Track->momentum().eta()) > trkEtaMax_)
```
Comment (sroychow) [inline L1Trigger/L1TTrackMatch/plugins/L1FastTrackingJetProducer.cc:161]: std::abs

File: L1Trigger/L1TTrackMatch/plugins/L1FastTrackingJetProducer.cc [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    iEvent.put(std::move(L1FastTrackingJets), "L1FastTrackingJetsExtended");
+  else
+    iEvent.put(std::move(L1FastTrackingJets), "L1FastTrackingJets");
+}
+
+void L1FastTrackingJetProducer::beginJob() {}
+
+void L1FastTrackingJetProducer::endJob() {}
+
+void L1FastTrackingJetProducer::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
```
Comment (sroychow) [inline L1Trigger/L1TTrackMatch/plugins/L1FastTrackingJetProducer.cc:248]: Suggest dropping the cfi file in favour of adding the fillDescriptions class. You can generate the pset description using 
```
edmPythonConfigToCppValidation L1FastTrackingJetProducer_cfi.py 
```

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
I have applied most of the comments except:
- I dont know how to solve the remaining static checks warnings
- I think it is not necessary to switch to range based loops everywhere
- edmPythonConfigToCppValidation L1FastTrackingJetProducer_cfi.py crashes and does not provide any output

**Comment by qliphy** [comment]
>     * edmPythonConfigToCppValidation L1FastTrackingJetProducer_cfi.py crashes and does not provide any output

If you remove 
https://github.com/cms-sw/cmssw/blob/eb7733a0703c4e0e81eb702620d123d7bee0689d/L1Trigger/L1TTrackMatch/python/L1FastTrackingJetProducer_cfi.py#L2

and then further change
https://github.com/cms-sw/cmssw/blob/eb7733a0703c4e0e81eb702620d123d7bee0689d/L1Trigger/L1TTrackMatch/python/L1FastTrackingJetProducer_cfi.py#L6
and
https://github.com/cms-sw/cmssw/blob/eb7733a0703c4e0e81eb702620d123d7bee0689d/L1Trigger/L1TTrackMatch/python/L1FastTrackingJetProducer_cfi.py#L28
to e.g.  simply " l1VertexCollectionName = cms.string("l1vertices"),"

then it would work with "edmPythonConfigToCppValidation L1FastTrackingJetProducer_cfi"

you can then also set up a test example file to load L1FastTrackingJetProducer_cfi and set L1PrimaryVertexTag to VertexProducer.l1VertexCollectionName.value()

**Comment by qliphy** [comment]
> Hi @cecilecaillol
> 
> Not related to this PR, but because it compiles L1Trigger/L1TTrackMatch, there are several warning I assume it should be checked/fixed.
> 
 
@cms-sw/core-l2   Do you have any suggestions on the static check warnings?

**Comment by qliphy** [comment]
> > Hi @cecilecaillol
> > Not related to this PR, but because it compiles L1Trigger/L1TTrackMatch, there are several warning I assume it should be checked/fixed.
> 
> @cms-sw/core-l2 Do you have any suggestions on the static check warnings?

https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c3465f/16272/llvm-analysis/runStaticChecks.log

**Comment by makortel** [comment]
Following the "Static checks outputs  See Static Checks" link gives an easier-to-digest report from the static analyzer that shows in the code the assumptions the static analyzer does to reach its conclusion.
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c3465f/16272/llvm-analysis/

Looking e.g. the first one
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-c3465f/16272/llvm-analysis/report-ba207a.html#EndPath
the problem the static analyzer sees is that this array
https://github.com/cms-sw/cmssw/blob/eb7733a0703c4e0e81eb702620d123d7bee0689d/L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc#L554-L555
is left uninitialized if `left_pt <= 0`
https://github.com/cms-sw/cmssw/blob/eb7733a0703c4e0e81eb702620d123d7bee0689d/L1Trigger/L1TTrackMatch/plugins/L1TrackJetProducer.cc#L584-L588

**Comment by cecilecaillol** [comment]
please test

**Comment by cecilecaillol** [comment]
I have added the fillDescriptions but not solved the static warnings. Would it be possible to merge this PR as is and @gkaratha, who is the original author of the code, would follow up with a PR to address the static warnings?

**Comment by cecilecaillol** [comment]
+l1

**Comment by srimanob** [comment]
> I have added the fillDescriptions but not solved the static warnings. Would it be possible to merge this PR as is and @gkaratha, who is the original author of the code, would follow up with a PR to address the static warnings?

Agree. The fix is beyond this PR, as noted in the original comment.

**Comment by qliphy** [comment]
@sroychow @srimanob   Do you have any further comments/suggestions on this PR?

**Comment by srimanob** [comment]
Hi @qliphy 
I am happy to sign. I would let @sroychow confirm if all fixed are OK. Thanks.

**Comment by sroychow** [comment]
@qliphy the PR is fine from my side. Thanks.

**Comment by srimanob** [comment]
+Upgrade

This PR is to add the FastTracking Jets by running FastJet clustering algorithm on L1 tracks.

**Comment by qliphy** [comment]
+1

**Pull request merged**